### PR TITLE
Use PARALLEL_FOR_IF with threadSafe function

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This adds a variadic function `ThreadSafe(...)` to determine whether one or more workspaces thread-safe. We can then remove  `PARALLEL_FOR1`, `PARALLEL_FOR2`, and `PARALLEL_FOR3` macros.

**To test:**

There is no GitHub issue associated with this pull request. 

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by quantumsteve** [comment]
@martyngigg I hid the function name inside `PARALLEL_FOR_IF(...)` and am debating whether this is better or worse than a macro-only solution. Does access to the type system justify the slightly uglier syntax?

**Comment by martyngigg** [comment]
@quantumsteve I don't know if it does. I think you first version, with the `threadSafe` part also appearing in the call to `PARALLEL_FOR_IF`, reads better as then you can see that the `if` is dependent on thread-safety rather than having to dig into the macro definition.

It would certainly be nice if `threadSafe` could accept references but `.get` is not horrendous.

File: Framework/Algorithms/inc/MantidAlgorithms/SpectrumAlgorithm.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ class MANTID_ALGORITHMS_DLL SpectrumAlgorithm : public API::Algorithm {
     auto size = static_cast<int64_t>(indexSet.size());
     API::Progress progress(this, 0.0, 1.0, size);
 
-    PARALLEL_FOR1((&workspace))
+    PARALLEL_FOR_IF(Kernel::threadSafe(workspace))
```
Comment (martyngigg) [inline Framework/Algorithms/inc/MantidAlgorithms/SpectrumAlgorithm.h]: I like this. It's nice that it now conveys that there is actually a `if` going on and it also describes the check too.

File: Framework/Algorithms/src/BinaryOperation.cpp [hunk 2]
```diff
@@ -492,7 +492,7 @@ void BinaryOperation::doSingleValue() {
 
   if (m_eout) {
     // ---- The output is an EventWorkspace ------
-    PARALLEL_FOR3(m_lhs, m_rhs, m_out)
+    PARALLEL_FOR_IF(Kernel::threadSafe(m_lhs.get(), m_rhs.get(), m_out.get()))
```
Comment (martyngigg) [inline Framework/Algorithms/src/BinaryOperation.cpp]: The single parameter variant now uses a reference. Is there a reason why the 3 parameter version has to call `.get()`?

Comment (quantumsteve) [inline Framework/Algorithms/src/BinaryOperation.cpp]: One should be able to use any combination of pointers and references. Pointers are checked for  `nullptr`, while references must never be NULL.

I changed the functions that were generating warnings with GCC 6.  Some of the three parameter versions can probably also be changed.

File: Framework/Algorithms/src/Rebin.cpp [hunk 3]
```diff
@@ -179,7 +179,7 @@ void Rebin::exec() {
       Progress prog(this, 0.0, 1.0, histnumber);
 
       // Go through all the histograms and set the data
-      PARALLEL_FOR3(inputWS, eventInputWS, outputWS)
+      PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *outputWS))
       for (int i = 0; i < histnumber; ++i) {
```
Comment (quantumsteve) [inline Framework/Algorithms/src/Rebin.cpp]: I removed `eventInputWS` since it points to the same workspace as `inputWS`.

**Comment by martyngigg** [comment]
@quantumsteve Is this ready to go now?

**Comment by quantumsteve** [comment]
@martyngigg Yes, I think so.

File: Framework/Algorithms/src/Q1D2.cpp [hunk 4]
```diff
@@ -146,7 +146,7 @@ void Q1D2::exec() {
   Progress progress(this, 0.05, 1.0, numSpec + 1);
 
   const auto &spectrumInfo = m_dataWS->spectrumInfo();
-  PARALLEL_FOR3(m_dataWS, outputWS, pixelAdj)
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_dataWS, *outputWS, pixelAdj.get()))
```
Comment (martyngigg) [inline Framework/Algorithms/src/Q1D2.cpp]: Should this be `*pixelAdj`?

Comment (quantumsteve) [inline Framework/Algorithms/src/Q1D2.cpp]: No, `pixelAdj` can be `NULL`.  I tried dereferencing it and the unit test failed.

Comment (martyngigg) [inline Framework/Algorithms/src/Q1D2.cpp]: I see, that make sense.

**Review by martyngigg — APPROVED** [review]
I'm happy with this now. It's a definite improvement in the readability and usability of these macros.

:shipit:

**Pull request merged**